File: BEMLIB_22.10/01_num_meth/03_lin_eq/pois_fds_DNDD_dr.f
      program pois_fds_DNDD_dr

c-----------------------------------------
c FDLIB
c
c Copyright by C. Pozrikidis, 1999
c All rights reserved.
c
c This program is to be used only under the
c stipulations of the licensing agreement.
c----------------------------------------

c------------------------------------------------
c This program accompanies the book:
c             C. Pozrikidis
c Numerical Computation in Science and Engineering
c        Oxford University Press
c------------------------------------------------

c-----------------------------------------
c Driver for generating a linear system
c for the finite-difference solution
c of the Poisson equation lapl(f)+g = 0,
c as discussed in the text
c-----------------------------------------

      Implicit Double Precision (a-h,o-z)

      Double Precision MAT(512,512),rhs(512)

      Dimension g(0:128,0:128)
      Dimension w(512),q(512),z(512),v(512)

c---
c input
c---

      write (6,*)
      write (6,*) " Please enter the box dimensions ax, bx, ay, by"
      write (6,*) " ----------------------------------------------"
      read  (5,*) ax,bx,ay,by

      write (6,*)
      write (6,*) " Please enter the discretization levels Nx and Ny"
      write (6,*) " ----------------------------------------------"
      read  (5,*) Nx,Ny

c--------------------
c boundary conditions
c and source term
c--------------------

      Do j=1,Ny+1
       w(j) = 0.0D0   ! example
       q(j) = 1.0D0   ! example
      End Do

      Do i=1,Nx+1
       z(i) = 0.0D0   ! example
       v(i) = 0.0D0   ! example
      End Do

      Do i=1,Nx+1
       Do j=1,Ny+1
        g(i,j) = 0.0D0  ! example
       End Do
      End Do

c---------------------------
c generate the linear system
c---------------------------

      call pois_fds_DNDD 
     +
     +   (ax,bx
     +   ,ay,by
     +   ,Nx
     +   ,Ny
     +   ,g
     +   ,w,q,z,v
     +   ,MAT,Rhs
     +   ,mats
     +   )

c-----------------
c printing session
c-----------------

      write (6,*)
      write (6,*)  "Linear system:"
      write (6,*)

      Do i=1,mats
       write (6,100) (MAT(i,j),j=1,mats),rhs(i)
      End Do

c-----
c done
c-----

  100 Format (100(1x,f6.3))

      Stop
      End

File: BEMLIB_22.10/01_num_meth/03_lin_eq/pois_fds_DNDD_dr.m
clear all
close all

%==========================================
% FDLIB
%
% Copyright by C. Pozrikidis, 1999
% All rights reserved.
%
% This program is to be used only under the
% stipulations of the licensing agreement
%==========================================

%------------------------------------------------
% This program accompanies the book:
%             C. Pozrikidis
% Numerical Computation in Science and Engineering
%        Oxford University Press
%------------------------------------------------

%-----------------------------------------
% Driver for generating a linear system
% for the finite-difference solution
% of the Poisson equation lapl(f)+g = 0,
% as discussed in the text
%-----------------------------------------

%---
% input
%---

 ax = 0.0;
 bx = 1.0;

 ay = 0.0;
 by = 1.0;

 Nx = 16;
 Ny = 16;

%---
% prepare
%---

 Dx = (bx-ax)/Nx;
 Dy = (by-ay)/Ny;


%--------------------
% boundary conditions
%--------------------

  for j=1:Ny+1
    w(j) = 0.0D0;  % example
    q(j) =-5.0D0;  % example
  end

  for i=1:Nx+1
    z(i) = 0.5D0;  % example
    v(i) =-1.0D0;  % example
  end

%---
% source
%---

  for j=1:Ny+1
    for i=1:Nx+1
      g(i,j) = 0.0D0; % example
    end
  end

%---------------------------
% generate the linear system
%---------------------------

  [mats,mat,rhs] = pois_fds_DNDD  ...
             ...
   (ax,bx ...
   ,ay,by ...
   ,Nx,Ny ...
   ,g ...
   ,w,q,z,v ...
   );

%-----------------
% printing session
%-----------------

%    mat


%---
% solution
%---

  sol = rhs/mat';


%---
% distribute the solution
%---

 p = 0;     % counter

  for j=2:Ny
   for i=2:Nx+1
     p = p+1;
     f(i,j) = sol(p);
   end
  end

  for j=1:Ny+1
    f(   1,j) = w(j);
  end

  for i=1:Nx+1
    f(i,   1) = z(i);
    f(i,Ny+1) = v(i);
  end

%---
% grid nodes
%---

for j=1:Ny+1
   for i=1:Nx+1
     x(i,j) = ax + (i-1)*Dx;
     y(i,j) = ay + (j-1)*Dy;
   end
end

%---
% plot
%---

%mesh(x,y,f)
surf(x,y,f)
%axis([0 2 0 1 0 0.1])
axis equal
box on
xlabel('x','fontsize',15)
ylabel('y','fontsize',15)
zlabel('f','fontsize',15)
set(gca,'fontsize',15)
view(-32,32)



